src/image-tvg.c

```diff
@@ -0,0 +1,1449 @@
+                                          int kind,
+                                          tvg_fill_header_t *out_header)
+{
+    uint32_t u32;
+    tvg_result_t res = tvg_read_varuint(ctx, &u32);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    size_t count = (size_t) u32 + 1;
+    out_header->size = count;
+    res = tvg_parse_style(ctx, kind, &out_header->style);
+    if (res != TVG_SUCCESS) {
+        log_error("Failed to parse style, error = [%d]", res);
+    }
+    return res;
+}
+
+static tvg_result_t tvg_parse_line_header(tvg_context_t *ctx,
+                                          int kind,
+                                          tvg_line_header_t *out_header)
+{
+    uint32_t u32;
+    tvg_result_t res = tvg_read_varuint(ctx, &u32);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    size_t count = (size_t) u32 + 1;
+    out_header->size = count;
+    res = tvg_parse_style(ctx, kind, &out_header->style);
+    if (res != TVG_SUCCESS) {
+        log_error("Failed to parse style, error = [%d]", res);
+        return res;
+    }
+    res = tvg_read_unit(ctx, &out_header->line_width);
+    if (res != TVG_SUCCESS) {
+        log_error("Failed to read unit, error = [%d]", res);
+        return res;
+    }
+
+    return TVG_SUCCESS;
+}
+
+static tvg_result_t tvg_parse_line_fill_header(
+    tvg_context_t *ctx,
+    int kind,
+    tvg_line_fill_header_t *out_header)
+{
+    uint8_t d;
+    if (1 > ctx->in(&d, 1, ctx->in_state)) {
+        return TVG_E_IO_ERROR;
+    }
+    tvg_result_t res = TVG_SUCCESS;
+
+    size_t count = TVG_SIZE_AND_STYLE_SIZE(d);
+    out_header->size = count;
+    res = tvg_parse_style(ctx, kind, &out_header->fill_style);
+    if (res != TVG_SUCCESS) {
+        log_error("Failed to parse style, error = [%d]", res);
+        return res;
+    }
+    res = tvg_parse_style(ctx, TVG_SIZE_AND_STYLE_STYLE_KIND(d),
+                          &out_header->line_style);
+    if (res != TVG_SUCCESS) {
+        log_error("Failed to parse style, error = [%d]", res);
+        return res;
+    }
+    res = tvg_read_unit(ctx, &out_header->line_width);
+    if (res != TVG_SUCCESS) {
+        log_error("Failed to read unit, error = [%d]", res);
+        return res;
+    }
+
+    return TVG_SUCCESS;
+}
+
+static tvg_result_t tvg_parse_path(tvg_context_t *ctx, size_t size)
+{
+    tvg_result_t res = TVG_SUCCESS;
+    tvg_point_t st, cur, pt;
+    float f32;
+    uint8_t path_info;
+    twin_path_t *path = ctx->path;
+    res = tvg_read_point(ctx, &pt);
+    if (res != TVG_SUCCESS) {
+        goto error;
+    }
+    twin_path_move(path, D(pt.x), D(pt.y));
+    st = pt;
+    cur = pt;
+    for (size_t j = 0; j < size; ++j) {
+        if (1 > ctx->in(&path_info, 1, ctx->in_state)) {
+            goto error;
+        }
+        float line_width = 0.0f;
+        if (TVG_PATH_CMD_HAS_LINE(path_info)) {
+            res = tvg_read_unit(ctx, &line_width);
+            if (res != TVG_SUCCESS) {
```

Comment:
I've noticed the code below is repeated in several places. Should it be refactored into a function for easier maintenance?
```c
if (res != TVG_SUCCESS) {
    log_error("Failed to read point, error = [%d]", res);
    goto error;
}
```